File: client/src/main/java/io/xol/chunkstories/input/lwjgl3/Lwjgl3KeyBind.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.input.lwjgl3;

import static org.lwjgl.glfw.GLFW.GLFW_PRESS;
import static org.lwjgl.glfw.GLFW.glfwGetKey;

import io.xol.chunkstories.api.client.Client;
import io.xol.chunkstories.api.client.IngameClient;
import io.xol.chunkstories.api.input.KeyboardKeyInput;
import io.xol.chunkstories.api.util.Configuration;
import io.xol.chunkstories.input.Pollable;

/**
 * Describes a key assignated to some action
 */
public class Lwjgl3KeyBind extends Lwjgl3Input implements KeyboardKeyInput, Pollable {
	private Lwjgl3ClientInputsManager inputsManager;

	private int GLFW_key;
	private int defaultKey;

	private boolean isDown = false;
	boolean editable = true;
	boolean repeat = false;

	private Configuration.OptionInt option;

	Lwjgl3KeyBind(Lwjgl3ClientInputsManager inputsManager, String name, String defaultKeyName) {
		super(inputsManager, name);
		this.inputsManager = inputsManager;
		this.defaultKey = GLFWKeyIndexHelper.getGlfwKeyByName(defaultKeyName);
		this.GLFW_key = defaultKey;

		Client client = inputsManager.gameWindow.getClient();
		Configuration clientConfiguration = client.getConfiguration();

		option = clientConfiguration.new OptionInt("client.input.bind." + name, defaultKey);
		option.addHook(o -> {GLFW_key = o.getValue();});
		clientConfiguration.registerOption(option);
	}

	private int parse(String s) {
		try {
			return Integer.parseInt(s);
		} catch (NumberFormatException e) {
			return 0;
		}
	}

	/**
	 * Internal to the engine, should not be interfered with by external mods
	 * 
	 * @return
	 */
	int getLWJGL3xKey() {
		return GLFW_key;
	}

	/** Returns true if the key is pressed and we're either not ingame or there is no GUI overlay blocking gameplay input */
	@Override
	public boolean isPressed() {
		IngameClient ingameClient = inputsManager.gameWindow.getClient().getIngame();
		if (ingameClient != null)
			return isDown && ingameClient.getPlayer().hasFocus();
		return isDown;
	}

	/**
	 * When reloading from the config file (options changed)
	 */
	public void reload() {
		// doesn't do stuff, we have a hook on the option directly
	}

	@Override
	public void updateStatus() {
		isDown = glfwGetKey(im.gameWindow.getGlfwWindowHandle(), GLFW_key) == GLFW_PRESS;
	}

	/**
	 * Is this key bind editable in the controls
	 */
	public boolean isEditable() {
		return editable;
	}

	public void setEditable(boolean editable) {
		this.editable = editable;
	}
}

File: client/src/main/java/xyz/chunkstories/input/lwjgl3/Lwjgl3KeyBind.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.input.lwjgl3

import org.lwjgl.glfw.GLFW
import org.lwjgl.glfw.GLFW.GLFW_PRESS
import org.lwjgl.glfw.GLFW.glfwGetKey

import xyz.chunkstories.api.input.KeyboardKeyInput
import xyz.chunkstories.input.Pollable

/**
 * Describes a key assignated to some action
 */
class Lwjgl3KeyBind internal constructor(inputsManager: Lwjgl3ClientInputsManager, name: String, defaultKeyName: String, val hidden: Boolean, val repeat: Boolean) : Lwjgl3Input(inputsManager, name), KeyboardKeyInput, Pollable {

    /**
     * Internal to the engine, should not be interfered with by external mods
     *
     * @return
     */
    internal var lwjglKey: Int = 0
        private set
    private val defaultKey: Int = GLFWKeyIndexHelper.getGlfwKeyByName(defaultKeyName) ?: GLFW.GLFW_KEY_UNKNOWN

    private var isDown = false

    init {
        this.lwjglKey = defaultKey

        val client = inputsManager.gameWindow.client
        val clientConfiguration = client.configuration

        if (!hidden) {
            val option = clientConfiguration.OptionKeyBind("client.input.bind.$name", defaultKey)
            option.addHook { lwjglKey = this.value }
            clientConfiguration.registerOption(option)
        }
    }

    /** Returns true if the key is pressed and we're either not ingame or there is no GUI overlay blocking gameplay input  */
    override val isPressed: Boolean
        get(): Boolean {
            val ingameClient = inputsManager.gameWindow.client.ingame
            return if (ingameClient != null) isDown && ingameClient.player.hasFocus() else isDown
        }


    override fun updateStatus() {
        isDown = glfwGetKey(inputsManager.gameWindow.glfwWindowHandle, lwjglKey) == GLFW_PRESS
    }
}
